server/src/termdb.server.init.js

```diff
@@ -106,33 +112,43 @@ export function server_init_db_queries(ds) {
 			return undefined
 		}
 	}
-
+	ds.term2SampleType = new Map()
+	if (ds.cohort.db.tableColumns['terms'].includes('sample_type')) {
+		let rows = cn.prepare('SELECT id, sample_type FROM terms').all()
+		for (const { id, sample_type } of rows) ds.term2SampleType.set(id, sample_type || DEFAULT_SAMPLE_TYPE)
+	}
 	if (tables.has('sampleidmap')) {
 		const i2s = new Map(),
 			s2i = new Map()
-		const s = cn.prepare('SELECT * FROM sampleidmap')
+		let rows = cn.prepare('SELECT * FROM sampleidmap').all()
 		let totalCount = 0
-		for (const { id, name, type } of s.all()) {
+		for (const { id, name } of rows) {
 			i2s.set(id, name)
 			s2i.set(name, id)
-			if (type == 'sample' || !type) totalCount++
+			totalCount++ //for dbs without cohorts or types
 		}
 		q.id2sampleName = id => i2s.get(id)
 		q.sampleName2id = s => s2i.get(s)
-
-		// if sampleidmap table is present, add this method to return sample count for datasets with and without subcohort
-		if (ds.cohort.termdb.selectCohort) {
-			// subcohort is enabled on this ds
-			const s = cn.prepare('SELECT cohort,sample_count from cohorts').all()
-			const h = {}
-			for (const i of s) h[i.cohort] = i.sample_count
-			q.getcohortsamplecount = cohortKey => {
-				if (!cohortKey) return totalCount
-				return h[cohortKey] || 0
+		if (tables.has('cohorts')) {
+			// if sampleidmap table is present, add this method to return sample count for datasets with and without subcohort
+			if (tables.has('cohort_sample_types')) rows = cn.prepare('SELECT * from cohort_sample_types').all()
+			else rows = cn.prepare('SELECT cohort,sample_count from cohorts').all()
+			q.getCohortSampleCount = cohortKey => {
+				let counts = rows
+					.filter(row => row.cohort == cohortKey)
+					.map(
+						row =>
+							`${row.sample_count} ${row.sample_type ? ds.types.get(row.sample_type) : 'sample'}${
+								row.sample_count > 1 ? 's' : ''
+							}`
+					)
+				let total = counts.join(' and ')
+				if (total == '')
+					//older db does not have types or sample_count
+					total = totalCount
```

Comment:
can use `total='n='+totalCount` so it doesn't just show a number for other ds

Reply:
I showed n samples, to keep the same format.